pinot-query-planner/src/main/java/org/apache/pinot/query/planner/DispatchablePlanFragment.java

```diff
@@ -0,0 +1,121 @@
+  private final List<WorkerMetadata> _workerMetadataList;
+
+  // This is used at broker stage - we don't need to ship it to the server.
+  private Map<QueryServerInstance, List<Integer>> _serverInstanceToWorkerIdMap;
+
+  // used for table scan stage - we use ServerInstance instead of VirtualServer
+  // here because all virtual servers that share a server instance will have the
+  // same segments on them
+  private Map<Integer, Map<String, List<String>>> _workerIdToSegmentsMap;
+  private Map<String, String> _customProperties;
+
+  public DispatchablePlanFragment(PlanFragment planFragment) {
+    this(planFragment, new ArrayList<>(), new HashMap<>(), new HashMap<>());
+  }
+
+  public DispatchablePlanFragment(PlanFragment planFragment, List<WorkerMetadata> workerMetadataList,
+      Map<QueryServerInstance, List<Integer>> serverInstanceToWorkerIdMap, Map<String, String> customPropertyMap) {
+    _planFragment = planFragment;
+    _workerMetadataList = workerMetadataList;
+    _serverInstanceToWorkerIdMap = serverInstanceToWorkerIdMap;
+    _customProperties = customPropertyMap;
+  }
+
+  public PlanFragment getPlanFragment() {
+    return _planFragment;
+  }
+
+  public List<WorkerMetadata> getWorkerMetadataList() {
+    return _workerMetadataList;
+  }
+
+  public Map<QueryServerInstance, List<Integer>> getServerInstanceToWorkerIdMap() {
+    return _serverInstanceToWorkerIdMap;
+  }
+
+  public Map<String, String> getCustomProperties() {
+    return _customProperties;
+  }
+
+  public String getTableName() {
+    return _customProperties.get(TABLE_NAME_KEY);
+  }
+
+  public String setTableName(String tableName) {
+    return _customProperties.put(TABLE_NAME_KEY, tableName);
+  }
+
+  public TimeBoundaryInfo getTimeBoundary() {
+    return new TimeBoundaryInfo(_customProperties.get(TIME_BOUNDARY_COLUMN_KEY),
+        _customProperties.get(TIME_BOUNDARY_VALUE_KEY));
+  }
+
+  public void setTimeBoundaryInfo(TimeBoundaryInfo timeBoundaryInfo) {
+    _customProperties.put(TIME_BOUNDARY_COLUMN_KEY, timeBoundaryInfo.getTimeColumn());
+    _customProperties.put(TIME_BOUNDARY_VALUE_KEY, timeBoundaryInfo.getTimeValue());
+  }
+
+  public Map<Integer, Map<String, List<String>>> getWorkerIdToSegmentsMap() {
+    return _workerIdToSegmentsMap;
+  }
+
+  public void setWorkerIdToSegmentsMap(Map<Integer, Map<String, List<String>>> workerIdToSegmentsMap) {
+    _workerIdToSegmentsMap = workerIdToSegmentsMap;
+  }
+
+  public List<String> getScannedTables() {
+    return _planFragment.getFragmentMetadata().getScannedTables();
+  }
+
+  public void setWorkerMetadataList(List<WorkerMetadata> workerMetadataList) {
+    _workerMetadataList.addAll(workerMetadataList);
+  }
+
+  public void setScannedTables(List<String> scannedTables) {
+    _planFragment.getFragmentMetadata().setScannedTables(scannedTables);
+  }
+
+  public PlanFragmentMetadata toPlanFragmentMetadata() {
+    return new PlanFragmentMetadata(_workerMetadataList, _customProperties);
+  }
```

Comment:
this constructor is confusing to me i don't see a PlanFragmentMetadata constructor that accepts the 2 arguments?

Reply:
This is the PlanFragmentMetadata in routing package, will refactor this to StageMetadata in another PR.

Reply:
Rename org.apache.pinot.query.routing.PlanFragmentMetadata to org.apache.pinot.query.routing.StageMetadata to distinguish.